File: cats/oaip/semester 2/D8 Судьба математика/solve.py
MAX = 1000000

#
#
# TL SOLVE
#
class Condition:
    def __init__(self, l, r, op):
        self.l = l
        self.r = r
        self.op = op


def to_cond(l):
    operand = l[1:-1]
    normal = {
        '=': 1,
        '>': 2,
        '<': 3,
        '<>': 4,
        '>=': 5,
        '<=': 6
    }
    html = {
        '&gt;': 2,
        '&lt;': 3,
        '&lt;&gt;': 4,
        '&gt;=': 5,
        '&lt;=': 6
    }
    try:
        temp = normal[operand]
    except:
        temp = html[operand]
    return Condition(int(l[0]), int(l[-1]), temp)


def is_right(j):
    global Conditions, Number
    op = Conditions[j].op
    l = Conditions[j].l
    r = Conditions[j].r
    if op == 1:
        return Number[l] == Number[r]
    elif op == 2:
        return Number[l] > Number[r]
    elif op == 3:
        return Number[l] < Number[r]
    elif op == 4:
        return Number[l] != Number[r]
    elif op == 5:
        return Number[l] >= Number[r]
    elif op == 6:
        return Number[l] <= Number[r]


def increment():
    global Number
    k = 6
    Number[k] += 1
    while Number[k] == 10:
        Number[k] = 0
        k -= 1
        Number[k] += 1


def main():
    global Conditions, Number
    with open('input.txt', 'r') as fin, open('output.txt', 'w') as fout:
        Conditions = [to_cond(line.strip()) for line in fin.readlines()]
        count = 0
        Number = [0] * 7
        for i in range(MAX):
            count += 1
            for j in range(len(Conditions)):
                if not is_right(j):
                    count -= 1
                    break
            increment()
        fout.write(str(count))


if __name__ == '__main__':
    main()

File: cats/oaip/semester 2/D8 Судьба математика/solve.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>

const int MAX = 1000000;

class Condition {
public:
    int l, r, op;

    Condition(int l, int r, int op) : l(l), r(r), op(op) {}
};

std::vector<Condition> Conditions;
std::vector<int> Number(7);

Condition to_cond(std::string l) {
    std::unordered_map<std::string, int> normal = {
            {"=",  1},
            {">",  2},
            {"<",  3},
            {"<>", 4},
            {">=", 5},
            {"<=", 6}
    };
    std::unordered_map<std::string, int> html = {
            {"&gt;",     2},
            {"&lt;",     3},
            {"&lt;&gt;", 4},
            {"&gt;=",    5},
            {"&lt;=",    6}
    };
    std::string operand = l.substr(1, l.size() - 2);
    int temp;
    try {
        temp = normal.at(operand);
    } catch (...) {
        temp = html.at(operand);
    }
    return Condition(l[0] - '0', l[l.size() - 1] - '0', temp);
}

bool is_right(int j) {
    int op = Conditions[j].op;
    int l = Conditions[j].l;
    int r = Conditions[j].r;
    switch (op) {
        case 1:
            return Number[l] == Number[r];
        case 2:
            return Number[l] > Number[r];
        case 3:
            return Number[l] < Number[r];
        case 4:
            return Number[l] != Number[r];
        case 5:
            return Number[l] >= Number[r];
        case 6:
            return Number[l] <= Number[r];
        default:
            return false;
    }
}

void increment() {
    int k = 6;
    Number[k]++;
    while (Number[k] == 10) {
        Number[k] = 0;
        k--;
        Number[k]++;
    }
}

int main() {
    std::ifstream fin("input.txt");
    std::ofstream fout("output.txt");

    std::string line;
    while (std::getline(fin, line)) {
        Conditions.push_back(to_cond(line));
    }

    int count = 0;
    Number.assign(7, 0);
    for (int i = 0; i < MAX; ++i) {
        count++;
        for (int j = 0; j < Conditions.size(); ++j) {
            if (!is_right(j)) {
                count--;
                break;
            }
        }
        increment();
    }

    fout << count;

    return 0;
}